### PR TITLE
Penalize large HTLCs relative to channels in default Scorer

### PR DESCRIPTION
**Pull request by TheBlueMatt**
Sending HTLCs which are any greater than a very small fraction of the
channel size tend to fail at a much higher rate. Thus, by default
we start applying a penalty at only 1/8th the channel size and
increase it linearly as the amount reaches the channel's capacity,
20 msat per 1024th of the channel capacity.

### CODE REVIEW

File: lightning/src/routing/mod.rs [hunk 1]
```diff
@@ -26,7 +26,14 @@ use sync::{Mutex, MutexGuard};
 pub trait Score {
 	/// Returns the fee in msats willing to be paid to avoid routing through the given channel
 	/// in the direction from `source` to `target`.
-	fn channel_penalty_msat(&self, short_channel_id: u64, source: &NodeId, target: &NodeId) -> u64;
+	///
+	/// The maximum amount which may be in the HTLC at the given channel is given by
```
Comment (jkczyz) [inline lightning/src/routing/mod.rs]: Sentence is a little hard to parse. Would it be simpler to drop it and instead update the previous sentence to include the variable? (i.e., Returns the fee in msats willing to be paid to avoid routing `send_amt_msat` through ...)

File: lightning/src/routing/mod.rs [hunk 2]
```diff
@@ -26,7 +26,14 @@ use sync::{Mutex, MutexGuard};
...
+	/// `send_amt_sat`.
+	///
+	/// The channel's capacity in satoshis (given other HTLCs to be sent) is given by
+	/// `channel_capacity_sats`. It may be guessed from various sources or assumed from no data.
+	/// Your code should be overflow-safe through a `channel_capacity_sats` of 21 million BTC.
+	fn channel_penalty_msat(&self, short_channel_id: u64, send_amt_msat: u64, channel_capacity_sat: u64, source: &NodeId, target: &NodeId) -> u64;
```
Comment (jkczyz) [inline lightning/src/routing/mod.rs]: Is this liquidity or capacity? Not sure if we've been consistent but I thought a channel has a total capacity with inbound and outbound liquidity being part of that. Or does each of those correspond to part of inbound and outbound capacity (i.e. accounting for inflight payments).

File: lightning/src/routing/scorer.rs [hunk 3]
```diff
@@ -98,16 +98,37 @@ pub struct ScorerUsingTime<T: Time> {
...
+	pub overuse_penalty_start_1024th: u16,
+
+	/// A penalty applied, per whole 1024ths of the channel value which the amount being sent over
+	/// the channel exceeds [`overuse_penalty_start_1024th`] by.
+	///
+	/// Default value: 20 msat (i.e. 2560 msat penalty to use 1/4th of a channel, 7680 msat penalty
+	///                to use half a channel, and 12,560 msat penalty to use 3/4ths of a channel)
+	///
+	/// [`overuse_penalty_start_1024th`]: Self::overuse_penalty_start_1024th
+	pub overuse_penalty_msat_per_1024th: u64,
```
Comment (jkczyz) [inline lightning/src/routing/scorer.rs:191]: Related to earlier comment, here we are saying "channel value". Should be consistent with how it is defined in `Score`.

File: lightning/src/routing/scorer.rs [hunk 4]
```diff
@@ -205,19 +230,29 @@ impl Default for ScoringParameters {
...
-		&self, short_channel_id: u64, _send_amt_msat: u64, _chan_capacity_sat: u64, _source: &NodeId, _target: &NodeId
+		&self, short_channel_id: u64, send_amt_msat: u64, chan_capacity_sat: u64, _source: &NodeId, _target: &NodeId
 	) -> u64 {
 		let failure_penalty_msat = self.channel_failures
 			.get(&short_channel_id)
 			.map_or(0, |value| value.decayed_penalty_msat(self.params.failure_penalty_half_life));
 
-		self.params.base_penalty_msat + failure_penalty_msat
+		let send_1024ths = send_amt_msat.checked_mul(1024).unwrap_or(u64::max_value())
+			/ (chan_capacity_sat.checked_mul(1000).unwrap_or(1));
```
Comment (jkczyz) [inline lightning/src/routing/scorer.rs:309]: Why not use `unwrap_or(u64::max_value())` here to indicate a large channel capacity?

**Review by jkczyz — COMMENTED** [review]
> Further, its incredibly awkward to have a `scorer`
> sub-module, but only have a single struct in it, with the relevant
> trait it is the only implementation of somewhere else.

I think the idea was to separate an implementation from its interface. Someone creating their own implementation shouldn't need to care about an implementation that we provide.

> Not having
> `Score` in the `scorer` sub-module is further confusing because
> it's the only module anywhere that references scoring at all.

It's used in the `router` module, so wouldn't that be a better place if we want to move it?  At very least it would be more consistent with other moves that we've made.

File: lightning/src/routing/mod.rs (hunk 2 above)
Comment (TheBlueMatt) [reply lightning/src/routing/mod.rs]: Capacity, we cannot know liquidity of a third-party channel.

File: lightning/src/routing/scorer.rs (hunk 4 above)
Comment (TheBlueMatt) [reply lightning/src/routing/scorer.rs:309]: I'd think we'd want to default to penalizing the channel - the channel is obviously bogus if we can't multiply its sat value by 1000, so we shouldn't divide by much.

**Comment by TheBlueMatt** [comment]
> I think the idea was to separate an implementation from its interface. Someone creating their own implementation shouldn't need to care about an implementation that we provide.

Right, I get that, but it just feels confusing for a user if you're importing two things that are obviously related from different modules, not to mention that it just generally makes it hard to work on this stuff - instead of having related code together we have several tiny files that we have to go edit any time we want to change something.

> It's used in the router module, so wouldn't that be a better place if we want to move it? At very least it would be more consistent with other moves that we've made.

I'd buy that, I guess. I still prefer keeping it with the implementation of it given just how small the files are and `router.rs` is already huge. In general I also think as we move towards more bindings users we need to rethink the whole "use this trait as module::Trait" thing because almost no users actually see that, and it results in traits in `mod.rs` which I've always found confusing.

File: lightning/src/routing/scorer.rs (hunk 4 above)
Comment (jkczyz) [reply lightning/src/routing/scorer.rs:305]: Would it make sense to return an infinite penalty in that case?

```rust
let chan_capacity_msat = match chan_capacity_sat.checked_mul(1000) {
    Some(chan_capacity_msat) => chan_capacity_msat,
    None => return u64::max_value(),
}
```

File: lightning/src/routing/scorer.rs [hunk 5]
```diff
@@ -244,8 +244,10 @@ impl<T: Time> routing::Score for ScorerUsingTime<T> {
 			.get(&short_channel_id)
 			.map_or(0, |value| value.decayed_penalty_msat(self.params.failure_penalty_half_life));
 
-		let send_1024ths = send_amt_msat.checked_mul(1024).unwrap_or(u64::max_value())
-			/ (chan_capacity_sat.checked_mul(1000).unwrap_or(1));
+		let chan_capacity_msat = if let Some(cap) = match chan_capacity_sat.checked_mul(1000)
+			{ cap } else { return u64::max_value() };
```
Comment (jkczyz) [inline lightning/src/routing/scorer.rs:305]: Looks like a compilation error here.

**Review by jkczyz — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
Squashed without diff to get ahead for other reviewers:

```
$ git diff-tree -U1 979f4801b c80ea9f87
$
```

File: lightning/src/routing/mod.rs [hunk 6]
```diff
@@ -24,9 +24,13 @@ use sync::{Mutex, MutexGuard};
...
 ///	Scoring is in terms of fees willing to be paid in order to avoid routing through a channel.
 pub trait Score {
-	/// Returns the fee in msats willing to be paid to avoid routing through the given channel
-	/// in the direction from `source` to `target`.
-	fn channel_penalty_msat(&self, short_channel_id: u64, source: &NodeId, target: &NodeId) -> u64;
+	/// Returns the fee in msats willing to be paid to avoid routing `send_amt_msat` through the
+	/// given channel in the direction from `source` to `target`.
+	///
+	/// The channel's capacity in satoshis (given other HTLCs to be sent) is given by
+	/// `channel_capacity_sats`. It may be guessed from various sources or assumed from no data.
```
Comment (valentinewallace) [inline lightning/src/routing/mod.rs]: Don't quite get the part "(given other HTLCs to be sent)". I think the first sentence can be removed?

File: lightning/src/routing/router.rs [hunk 7]
```diff
@@ -893,7 +893,7 @@ where L::Target: Logger {
 							}
 
 							let path_penalty_msat = $next_hops_path_penalty_msat
-								.checked_add(scorer.channel_penalty_msat($chan_id.clone(), &$src_node_id, &$dest_node_id))
+								.checked_add(scorer.channel_penalty_msat($chan_id.clone(), amount_to_transfer_over_msat, *available_liquidity_msat, &$src_node_id, &$dest_node_id))
```
Comment (valentinewallace) [inline lightning/src/routing/router.rs:896]: I think the channel capacity is supposed to be in `satoshis` not `msat`? Similar below

Comment (TheBlueMatt) [reply lightning/src/routing/router.rs]: I updated this and also changed the last-hop behavior to be completely different - `u64::max_value()`, cause, like, its really "unlimited value" as far as we're concerned.

**Comment by renepickhardt** [comment]
I am sure you have seen and studied the recent improvements to c-lightning. They actually penalized smaller channels by computing the negative log of the exact channel success probability (assuming a uniform distribution which we have widly seen on the network.)

In this comment https://github.com/ElementsProject/lightning/pull/4771#issuecomment-925145224 I explain that the linerized version (first term of the Taylor approximation) of the negative log prob is `amt/cap`. This es btw exactly the failure probability. However mainnet tests (and simulations) indicate that the actual log prob  works better than the linearized term. This I don't understand your (in my opinion) arbitrary chosen start at 1/8th of the capacity. Why not switching to the full probabilistic model instead? 

Since you don't compute min cost flows you don't need convex and you could even have concave cost functions (which is why we introduced the harmonic mean to compute an average of all features)

File: lightning/src/routing/router.rs [hunk 8]
```diff
@@ -893,7 +893,7 @@ where L::Target: Logger {
 							}
 
 							let path_penalty_msat = $next_hops_path_penalty_msat
-								.checked_add(scorer.channel_penalty_msat($chan_id.clone(), &$src_node_id, &$dest_node_id))
+								.checked_add(scorer.channel_penalty_msat($chan_id.clone(), amount_to_transfer_over_msat, *available_liquidity_msat / 1000, &$src_node_id, &$dest_node_id))
```
Comment (jkczyz) [inline lightning/src/routing/router.rs:896]: Any reason to not use `msats` for channel capacity in `channel_penalty_msat`'s interface? You are multiplying by 1000 in the implementation already.

**Comment by TheBlueMatt** [comment]
> I am sure you have seen and studied the recent improvements to c-lightning

I had seen it mentioned, but had not had time to dig into it in any detail at all, thanks for your comments!

> In this comment ElementsProject/lightning#4771 (comment) I explain that the linerized version (first term of the Taylor approximation) of the negative log prob is amt/cap. This es btw exactly the failure probability. However mainnet tests (and simulations) indicate that the actual log prob works better than the linearized term. This I don't understand your (in my opinion) arbitrary chosen start at 1/8th of the capacity. Why not switching to the full probabilistic model instead?

Good to know! We should probably look into using a log term here then. Indeed, the constants picked here for default are relatively arbitrary, based largely on intuition and not having found the time to dig into your research in depth to find the "right" numbers here, just getting the API right so we can tune it later.

For now I'd prefer to keep it linear, as log calls can sometimes be pretty slow and I want to do some additional tweaks to the router to tune performance and improve the path-combining pass before we start adding things that may slow it down.

File: lightning/src/routing/router.rs (hunk 8 above)
Comment (TheBlueMatt) [reply lightning/src/routing/router.rs:896]: I guess not? felt weird cause a channel's "capacity" is never msat, but I suppose we do use `htlc_maximum_msat` without issue. Will change.

File: lightning/src/routing/mod.rs [hunk 9]
```diff
@@ -24,9 +24,18 @@ use sync::{Mutex, MutexGuard};
...
-	fn channel_penalty_msat(&self, short_channel_id: u64, source: &NodeId, target: &NodeId) -> u64;
+	/// Returns the fee in msats willing to be paid to avoid routing `send_amt_msat` through the
+	/// given channel in the direction from `source` to `target`.
+	///
+	/// The channel's capacity (less any other MPP parts which are also being considered for use in
+	/// the same payment) is given by `channel_capacity_msat`. It may be guessed from various
+	/// sources or assumed from no data at all.
+	///
+	/// For hints provided in the invoice, we assume the channel has sufficient capacity to accept
+	/// the invoice's full amount, and provide a `channel_capacity_sat` of `u64::max_value()`.
```
Comment (jkczyz) [inline lightning/src/routing/mod.rs]: s/channel_capacity_sat/channel_capacity_msat

File: lightning/src/routing/mod.rs [hunk 10]
```diff
@@ -65,8 +74,8 @@ impl<'a, T: 'a + Score> LockableScore<'a> for RefCell<T> {
 }
 
 impl<S: Score, T: DerefMut<Target=S>> Score for T {
-	fn channel_penalty_msat(&self, short_channel_id: u64, source: &NodeId, target: &NodeId) -> u64 {
-		self.deref().channel_penalty_msat(short_channel_id, source, target)
+	fn channel_penalty_msat(&self, short_channel_id: u64, send_amt_msat: u64, channel_capacity_sat: u64, source: &NodeId, target: &NodeId) -> u64 {
+		self.deref().channel_penalty_msat(short_channel_id, send_amt_msat, channel_capacity_sat, source, target)
```
Comment (jkczyz) [inline lightning/src/routing/mod.rs]: Likewise.

**Review by jkczyz — APPROVED** [review]

**Review by valentinewallace — COMMENTED** [review]
It seems weird to assume that each route hint has sufficient capacity for the whole payment, just because I'd think users with small channels might want to pool inbound capacity across their channels? 

Seems like a preexisting thing though, so I'm pretty much ACK

**Comment by TheBlueMatt** [comment]
> It seems weird to assume that each route hint has sufficient capacity for the whole payment, just because I'd think users with small channels might want to pool inbound capacity across their channels?

Yea, its not entirely clear what we should do here. I think the right answer is to try to split more in the combine-paths pass at the end instead of doing it during normal graph-walking, and always assume enough value is available. In any case, as you note, that isn't new here.

**Comment by valentinewallace** [comment]
> > It seems weird to assume that each route hint has sufficient capacity for the whole payment, just because I'd think users with small channels might want to pool inbound capacity across their channels?
> 
> Yea, its not entirely clear what we should do here. I think the right answer is to try to split more in the combine-paths pass at the end instead of doing it during normal graph-walking, and always assume enough value is available. In any case, as you note, that isn't new here.

Could utilize `htlc_maximum_msat` too if available

**Review by valentinewallace — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
> Could utilize htlc_maximum_msat too if available

It sadly, basically never is - there's no field for it in the BOLT 11 invoice.

**Review by jkczyz — APPROVED** [review]

**Comment by TheBlueMatt** [comment]
Squashed without change, will land after CI:

```
$ git diff-tree -U1 3d3060588 3f9adb277
$
```

**Comment by TheBlueMatt** [comment]
Oops, needed rebase, range diff is:

```diff
$ git range-diff 7c4dfad4fe97c9d8bc37c38e15b9dab95f27e37a...3f9adb277514dbe464d198e67a3c9fc96e292395 e1ad422c1b9bb114f20411777faaee750c670ec5...42ebf774155632b5656fd5820eb8c28d0003d9b6
1:  679bef2d8 = 1:  8dc7cfab3 Provide `Score` the HTLC amount and channel capacity
2:  092128279 = 2:  9bec35ddd Penalize large HTLCs relative to channels in default `Scorer`
3:  3f9adb277 ! 3:  42ebf7741 Move `Score` into a `scoring` module instead of a top-level module
    @@ fuzz/src/router.rs: use lightning::ln::channelmanager::{ChannelDetails, ChannelC
     
      ## lightning-invoice/src/payment.rs ##
     @@
    - //! # use lightning::ln::{PaymentHash, PaymentSecret};
    + //! # use lightning::ln::{PaymentHash, PaymentPreimage, PaymentSecret};
      //! # use lightning::ln::channelmanager::{ChannelDetails, PaymentId, PaymentSendFailure};
      //! # use lightning::ln::msgs::LightningError;
     -//! # use lightning::routing;
    @@ lightning-invoice/src/payment.rs
     -//! # impl<S: routing::Score> Router<S> for FakeRouter {
     +//! # impl<S: Score> Router<S> for FakeRouter {
      //! #     fn find_route(
    - //! #         &self, payer: &PublicKey, params: &RouteParameters,
    + //! #         &self, payer: &PublicKey, params: &RouteParameters, payment_hash: &PaymentHash,
      //! #         first_hops: Option<&[&ChannelDetails]>, scorer: &S
     @@
      //! # }
    @@ lightning-invoice/src/payment.rs
      //! #     fn channel_penalty_msat(
      //! #         &self, _short_channel_id: u64, _send_amt: u64, _chan_amt: Option<u64>, _source: &NodeId, _target: &NodeId
      //! #     ) -> u64 { 0 }
    -@@ lightning-invoice/src/payment.rs: use bitcoin_hashes::Hash;
    - use lightning::ln::{PaymentHash, PaymentSecret};
    +@@ lightning-invoice/src/payment.rs: use bitcoin_hashes::sha256::Hash as Sha256;
    + use lightning::ln::{PaymentHash, PaymentPreimage, PaymentSecret};
      use lightning::ln::channelmanager::{ChannelDetails, PaymentId, PaymentSendFailure};
      use lightning::ln::msgs::LightningError;
     -use lightning::routing;
    @@ lightning-invoice/src/payment.rs: pub trait Payer {
     +pub trait Router<S: Score> {
        /// Finds a [`Route`] between `payer` and `payee` for a payment with the given values.
        fn find_route(
    -           &self, payer: &PublicKey, params: &RouteParameters, first_hops: Option<&[&ChannelDetails]>,
    +           &self, payer: &PublicKey, params: &RouteParameters, payment_hash: &PaymentHash,
     @@ lightning-invoice/src/payment.rs: pub enum PaymentError {
      impl<P: Deref, R, S: Deref, L: Deref, E> InvoicePayer<P, R, S, L, E>
      where
    @@ lightning-invoice/src/payment.rs: mod tests {
     -  impl<S: routing::Score> Router<S> for TestRouter {
     +  impl<S: Score> Router<S> for TestRouter {
                fn find_route(
    -                   &self,
    -                   _payer: &PublicKey,
    +                   &self, _payer: &PublicKey, params: &RouteParameters, _payment_hash: &PaymentHash,
    +                   _first_hops: Option<&[&ChannelDetails]>, _scorer: &S
     @@ lightning-invoice/src/payment.rs: mod tests {
      
        struct FailingRouter;
    @@ lightning-invoice/src/payment.rs: mod tests {
     -  impl<S: routing::Score> Router<S> for FailingRouter {
     +  impl<S: Score> Router<S> for FailingRouter {
                fn find_route(
    -                   &self,
    -                   _payer: &PublicKey,
    +                   &self, _payer: &PublicKey, _params: &RouteParameters, _payment_hash: &PaymentHash,
    +                   _first_hops: Option<&[&ChannelDetails]>, _scorer: &S
     @@ lightning-invoice/src/payment.rs: mod tests {
                }
        }
    @@ lightning-invoice/src/payment.rs: mod tests {
      
     -  impl<S: routing::Score> Router<S> for ManualRouter {
     +  impl<S: Score> Router<S> for ManualRouter {
    -           fn find_route(&self, _payer: &PublicKey, _params: &RouteParameters, _first_hops: Option<&[&ChannelDetails]>, _scorer: &S)
    -           -> Result<Route, LightningError> {
    -                   self.0.borrow_mut().pop_front().unwrap()
    +           fn find_route(
    +                   &self, _payer: &PublicKey, _params: &RouteParameters, _payment_hash: &PaymentHash,
    +                   _first_hops: Option<&[&ChannelDetails]>, _scorer: &S
     
      ## lightning-invoice/src/utils.rs ##
     @@ lightning-invoice/src/utils.rs: use lightning::chain::keysinterface::{Sign, KeysInterface};
    - use lightning::ln::{PaymentHash, PaymentSecret};
    + use lightning::ln::{PaymentHash, PaymentPreimage, PaymentSecret};
      use lightning::ln::channelmanager::{ChannelDetails, ChannelManager, PaymentId, PaymentSendFailure, MIN_FINAL_CLTV_EXPIRY};
      use lightning::ln::msgs::LightningError;
     -use lightning::routing;
    @@ lightning-invoice/src/utils.rs: impl<G, L: Deref> DefaultRouter<G, L> where G: D
     +impl<G, L: Deref, S: Score> Router<S> for DefaultRouter<G, L>
      where G: Deref<Target = NetworkGraph>, L::Target: Logger {
        fn find_route(
    -           &self, payer: &PublicKey, params: &RouteParameters, first_hops: Option<&[&ChannelDetails]>,
    +           &self, payer: &PublicKey, params: &RouteParameters, _payment_hash: &PaymentHash,
     
      ## lightning/src/ln/channelmanager.rs ##
     @@ lightning/src/ln/channelmanager.rs: pub mod bench {
$
```

**Review by jkczyz — APPROVED** [review]
ACK 42ebf774155632b5656fd5820eb8c28d0003d9b6

**Comment by renepickhardt** [comment]
> Good to know! We should probably look into using a log term here then. Indeed, the constants picked here for default are relatively arbitrary, based largely on intuition and not having found the time to dig into your research in depth to find the "right" numbers here, just getting the API right so we can tune it later.

Makes sense to start with the api. just for your reference I added a few words to lnds current issue of why chosing the log term over arbitrary constants is preferable https://github.com/lightningnetwork/lnd/issues/5988#issuecomment-971478023

That being said I did extensive simulations with the linerized version as linear min cost flow solvers are much faster than convex ones. The issue (which in practice can be modeled away) is that the linearized version still tends to fully saturate channels, similar as optimizing just for routing fees) which we already know is not preferable 

> For now I'd prefer to keep it linear, as log calls can sometimes be pretty slow and I want to do some additional tweaks to the router to tune performance and improve the path-combining pass before we start adding things that may slow it down.

Sorry but that sounds like premature optimization. If speed at that point should ever be an issue you could just take the most significant bit as an approximation of the log which should be a CPU instruction and thus one clock cycle. Also note that for the end user perspective taking a less likely path means more attempts on average and the roundtrip times of onions are dominant (probably even only local signature computation for moving state is more expensive than computing ba few logs in your scoring function) 

Anyway let me know if you have questions or need help

**Comment by TheBlueMatt** [comment]
> Makes sense to start with the api. just for your reference I added a few words to lnds current issue of why chosing the log term over arbitrary constants is preferable lightningnetwork/lnd#5988 (comment)

Cool. Thanks, yea, really curious what results you get when looking at decay parameters and how to incorporate failure/success information. As you maybe saw here we currently do the most basic naive failure incorporation possible. We have an issue for "how do we do something smarter" at #1170 and its not actually clear to me how to incorporate the amount of the failed HTLC without making the scoring super complicated, any insight you might have there would be super helpful.

> The issue is that the linearized version still tends to fully saturate channels

Yep, that makes sense. We currently already tend to saturate channels cause we don't use a flow solver anyway. Another next step here is changing our "combine multiple paths into a single route" pass at the end of the router to consider this better and avoid saturating more than we need to.

> Sorry but that sounds like premature optimization

Maybe, but one issue is that a lot of our inner routing loop is inside a macro which in current rust does not expose well to profilers - it just all shows up as "you have CPU usage inside the macro". As you've noted we have a lot of next-steps here, so I'm not trying to push back and suggest we definitely shouldn't do something, only that we have so many steps to go that there's no need to be in a rush for any specific step, and because profiling doesn't work currently here we really need to make sure we have a good baseline before we start adding additional complexity. Before we can even consider the current code a baseline there's also a few more optimizations we want to land. Its all just a question of sequencing :).
